doc/1.4/language.md

```diff
@@ -1221,14 +1221,121 @@ template get_qname {
 }
 ```
 
-### Typed Parameters
-A *typed parameter declaration* is a parameter declaration form that may only
-appear within templates:
+## Parameters detailed
 
-<pre>
-param <em>name</em> : <em>type</em>;
-</pre>
+Parameters may be typed or untyped.
+Typed parameter declarations may only appear in template definitions.
+
+Parameters are declared using one of the forms
+
+```
+param name;
+param name: type
+```
+
+and assigned using the `=` operator. Parameters may also be given default values using the form `param name default expr`.
+For example:
+
+```
+param offset = 8;
+param byte_order default "little-endian";
+```
+
+A default value is overridden by an assignment (`=`).
+There can be at most one assignment for each parameter.
+Typically, a default value for a parameter is specified in a template, and the programmer may then choose to override it where the template is used.
+See [Resolution of overrides](#resolution-of-overrides) for the resolution order when there are multiple definitions of the same parameter.
+
+A parameter that is declared without an assignment or a default value must eventually be assigned elsewhere, or the model will not compile.
+This pattern is sometimes useful in templates, as in:
+
+```
+template constant is register {
+    param value;
+    method get() -> (uint64) {
+        return value;
+    }
+}
+```
+
+so that wherever the template `constant` is used, the programmer
+is also forced to define the parameter `value`. E.g.:
+
+```
+register r0 size 2 @ 0x0000 is (constant) {
+    param value = 0xffff;
+}
+```
+
+> [!IMPORTANT]
+> When writing templates, always declare parameters that are referenced.
+> Enabling the provisional feature `explicit_param_decls` enforces this.
+>
+> Leaving out the parameter declaration from the template
+> definition can have unwanted effects if the programmer forgets to
+> specify its value where the template is used. At best, it will only
+> cause a more obscure error message, such as "unknown identifier"; at
+> worst, the scoping rules will select an unrelated definition of the same
+> parameter name.
+
+There is a shorthand syntax for combined declaration and definition of a parameter, currently enabled by the [`explicit_param_decls` provisional feature](provisional-auto.html#explicit_param_decls):
+
+```
+param NAME: TYPE = value;
+param NAME: TYPE default value;
+param NAME := value;
+param :default value;
+```
+
+`explicit_param_decls` enforces that parameters are declared before they are assigned, or that the combined syntax is used.
+This distinguishes between the intent to declare a new parameter, and the intent to override an existing parameter.
+This distinction allows DML to capture misspelled parameter overrides as compile errors.
 
+DMLC signals an error if the combined declaration and definition syntax is used to override an existing parameter.
+This guards against unintentional reuse of a parameter name. An example:
```

Comment:
Consider if it gets better by mentioning the motivation for the provisional earlier, e.g. right before or after you introduce it.

Reply:
Yeah I disagree with Erik on this front. There's nothing wrong with presenting the motivation after describing the feature.